tracer/src/Datadog.Trace/Debugger/Symbols/SymbolExtractor.cs

```diff
@@ -0,0 +1,718 @@
+        protected const int UnknownFieldAndArgLine = 0;
+        private const MethodAttributes StaticFinalVirtual = MethodAttributes.Static | MethodAttributes.Final | MethodAttributes.Virtual;
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(SymbolExtractor));
+        private readonly Dictionary<int, string> _methodAccess = new()
+        {
+            { 0x0001, "Private" },
+            { 0x0002, "FamANDAssem" },
+            { 0x0003, "Assembly" },
+            { 0x0004, "Family" },
+            { 0x0005, "FamORAssem" },
+            { 0x0006, "Public" }
+        };
+
+        private readonly Dictionary<int, string> _methodAttributes = new()
+        {
+            { 0x0010, "Static" },
+            { 0x0020, "Final" },
+            { 0x0040, "Virtual" },
+            { 0x0060, "Final Virtual" },
+        };
+
+        private bool _disposed;
+
+        protected SymbolExtractor(DatadogMetadataReader metadataReader, string assemblyPath)
+        {
+            _assemblyPath = assemblyPath;
+            DatadogMetadataReader = metadataReader;
+            MetadataReader = metadataReader.MetadataReader;
+        }
+
+        protected DatadogMetadataReader DatadogMetadataReader { get; }
+
+        protected MetadataReader MetadataReader { get; }
+
+        public static SymbolExtractor? Create(Assembly assembly)
+        {
+            try
+            {
+                var datadogMetadataReader = DatadogMetadataReader.CreatePdbReader(assembly);
+                if (datadogMetadataReader.MetadataReader.TypeDefinitions.Count == 0)
+                {
+                    Log.Debug("Could not found any type in assembly {Assembly}", assembly.FullName);
+                    return null;
+                }
+
+                return datadogMetadataReader.IsPdbExist ? new SymbolPdbExtractor(datadogMetadataReader, assembly.Location) : new SymbolExtractor(datadogMetadataReader, assembly.Location);
+            }
+            catch (Exception e)
+            {
+                Log.Error(e, "Error while creating instance of SymbolExtractor with {Assembly}", assembly.FullName);
+                return null;
+            }
+        }
+
+        internal Model.Scope GetAssemblySymbol()
+        {
+            var assemblyNameHandle = MetadataReader.GetAssemblyDefinition().Name;
+            var assemblyName = assemblyNameHandle.IsNil ? null : MetadataReader.GetString(assemblyNameHandle);
+            var assemblyScope = new Model.Scope
+            {
+                Name = assemblyName,
+                ScopeType = SymbolType.Assembly,
+                SourceFile = string.IsNullOrEmpty(_assemblyPath) ? null : _assemblyPath,
+            };
+
+            return assemblyScope;
+        }
+
+        internal IEnumerable<Model.Scope?> GetClassSymbols()
+        {
+            foreach (var typeDefinitionHandle in MetadataReader.TypeDefinitions)
+            {
+                if (typeDefinitionHandle.IsNil)
```

Comment:
I tend to think we really need a try/catch here (which might mean this can't be an iterator method), as otherwise, failing to capture a single class may mean not having all the symbols for the entire assembly?

Reply:
The single entry method for extracting metadata info for class is `TryGetClassSymbols` which is has wrapped in a `try/catch` so if something wrong happened, we fail for this specific class but not for the entire assembly. If you concern about this expression `MetadataReader.TypeDefinitions`, my assumption that if this will fail, we should skip this assembly.
Is it make sense?

Reply:
Why try/catch at the class granularity instead of at the member (method/field) granularity? Seems like it'd be a shame to miss out an entire class just because of one 'problematic' member?

Reply:
Short answer: I'll wrap also methods with try\catch .

Long answer:
I don't have a definitive answer to that. Wrapping each member with try\catch will ensure that we lose as little information as possible. 
On the other hand is this information good for us when it is incomplete?
For example we rely on all the methods to know the class StatLine and EndLine. Or what about method that we can get his closures but not the method itself? do we want this partial info?
I think that we should handle all cases that we do want missing info but if an exception has been thrown it means that we didn't think in ahead about this case so we don't know if this misleading info will help us or confuse us.

Reply:
Fixed in d76bac0e9bd3cfbda3e480deff9bbd0e453a32ec